### PR TITLE
Emit WebAssembly by default

### PR DESCRIPTION
**Pull request by kripken**
This changes the default output from asm.js to WebAssembly. Only the default is changed - we of course still support asm.js, and will for a very long time. But changing the default makes sense as the recommended output for most use cases should be WebAssembly, given it has shipped in all major browsers and platforms and is more efficient than asm.js.

The main changes in this PR are:
 * The actual flipping of the switch to wasm.
 * Cleanups in emcc.py and emscripten.py for that flag change - I swapped all BINARYEN to WASM, which made reading the a code a lot easier, at least for me. I also added some clearer warnings/errors for mismatched flags (features we just support in asm.js, like split memory).
 * Update the test suite. A lot of tests required some changes, e.g. because some are actually asm.js tests (so we must force wasm off) or they assume sync startup (which wasm does not do by default), etc. This is a big change, but nothing really sophisticated there.
 * Update the tutorial, FAQ, optimization docs, etc.

Some tricky things:

 * I don't really understand how the wasm-pthreads-browser tests work. I rewrote that section so that it keeps the tests passing in the default mode (asm.js), but I'm not sure how to check the wasm mode, or what we want there (why do we have both `also_asmjs` as a property and a variable, and check both?) - cc @dschuff 
 * For configuration checks in cmake and `./configure` etc., I disabled wasm. This is much simpler, as without wasm we have just a single file, with sync startup, and don't depend on wasm support in the VM (e.g., maybe the VM has no wasm support and  the user will build to asm.js later anyhow). But, maybe there is some issue here I am missing.

Otherwise, the PR should be mostly straightforward. I split off all the trickier parts I could into separate PRs that have all landed already.

### CODE REVIEW

**Comment by kripken** [comment]
Oh, and also another notable change in this PR is

 * Make `LEGACY_VM_SUPPORT` flip off wasm. When targeting old VMs, IE11, etc., wasm doesn't make sense. As a result, this PR should not break code using that option.

**Comment by dschuff** [comment]
The reason we had `also_wasm` as a property and an argument is that the property is set from the environment (so you can use one flag to enable/disable testing everything with wasm in addition to asmjs in the same test run), and then as an argument as an easy to run every test the same way again (i.e. just make a recursive call with all the same flags except `also_wasm`). It's basically there because the browser suite doesn't have the different modes the way the core suite does. At one point I had a second env var for pthreads since the browsers have those off by default still, but I don't see it in there now. So the pthreads tests should be no different from the other tests.

File: tests/test_browser.py [hunk 1]
```diff
@@ -3014,7 +3027,7 @@ def test_emterpreter_async_sleep2_safeheap(self):
 
   @requires_hardware
   def test_sdl_audio_beep_sleep(self):
-    self.btest('sdl_audio_beep_sleep.cpp', '1', args=['-s', 'EMTERPRETIFY=1', '-s', 'EMTERPRETIFY_ASYNC=1', '-Os', '-s', 'ASSERTIONS=1', '-s', 'DISABLE_EXCEPTION_CATCHING=0', '-profiling', '-s', 'SAFE_HEAP=1', '-lSDL'], timeout=60)
+    self.btest('sdl_audio_beep_sleep.cpp', '1', args=['-s', 'EMTERPRETIFY=1', '-s', 'EMTERPRETIFY_ASYNC=1', '-Os', '-s', 'ASSERTIONS=1', '-s', 'DISABLE_EXCEPTION_CATCHING=0', '-profiling', '-s', 'SAFE_HEAP=1', '-lSDL'], timeout=90)
```
Comment (jgravelle-google) [inline tests/test_browser.py]: Does wasm make the test potentially take longer?

File: tests/test_browser.py [hunk 2]
```diff
@@ -3065,7 +3078,8 @@ def test_modularize(self):
         print('test on', opts, args, code)
         src = open(path_from_root('tests', 'browser_test_hello_world.c')).read()
         open('test.c', 'w').write(self.with_report_result(src))
-        Popen([PYTHON, EMCC, 'test.c', '-s', 'MODULARIZE=1'] + args + opts).communicate()
+        # this test is synchronous, so avoid async startup due to wasm features
+        Popen([PYTHON, EMCC, 'test.c', '-s', 'MODULARIZE=1', '-s', 'BINARYEN_ASYNC_COMPILATION=0', '-s', 'SINGLE_FILE=1'] + args + opts).communicate()
```
Comment (jgravelle-google) [inline tests/test_browser.py]: `-s WASM=1 -s BINARYEN_ASYNC_COMPILATION=0` is a bit weird. We should probably alias these... in some future PR this one's big enough

**Comment by kripken** [comment]
Thanks @dschuff , now I think I get it. See the last commit, that should refactor that to be simpler and achieve the same results,

 * Without that `EMCC_BROWSER_ALSO_ASMJS` env var, we just run browser tests normally (using wasm, unless a test says otherwise). But pthreads tests we run in asm.js since wasm isn't stable yet.
 * With that env var, we run the test also in asm.js mode (we don't need the extra `also_asmjs` arg to do that, can just look at the arguments, do they disable wasm or not).

File: tests/test_other.py [hunk 3]
```diff
@@ -361,10 +431,12 @@ def test_emcc_multiprocess_cache_access(self):
       tasks = []
       num_times_libc_was_built = 0
       for i in range(3):
-        p = subprocess.Popen([PYTHON, EMCC, c_file, '--cache', cache_dir_name], stderr=subprocess.STDOUT, stdout=PIPE, universal_newlines=True)
+        p = subprocess.Popen([PYTHON, EMCC, c_file, '--cache', cache_dir_name, '-o', '%d.js' % i], stderr=subprocess.STDOUT, stdout=PIPE, universal_newlines=True)
         tasks += [p]
       for p in tasks:
         stdout, stderr = p.communicate()
+        print('O\n', stdout)
```
Comment (jgravelle-google) [inline tests/test_other.py]: Maybe stdout/stderr instead of O/E?

File: tests/test_other.py [hunk 4]
```diff
@@ -3634,25 +3721,41 @@ def test_bad_function_pointer_cast(self):
...
+                        assert 'Invalid function pointer called' in output, output
+                      else:
+                        assert 'abort(' in output, output
+                  else:
+                    assert 'Function table mask error' in output, output
                 else:
-                  assert 'abort(' in output, output
+                  if opts == 0:
+                    if wasm and (relocate or emulate_fps):
+                      assert 'function signature mismatch' in output, output
```
Comment (jgravelle-google) [inline tests/test_other.py]: This is all a tad horrifying
Maybe something like:
```python
for emulate_casts, emulate_fps, relocate, wasm, expected in [
  (0, 0, 0, 0, 'Invalid function pointer called'),
  (0, 0, 0, 1, 'Invalid function pointer called'),
  ...
]:
  ...
  assert expected in output, 'Expected to see ' + expected + ' in ' + output
```
Generally flatten it out and make it clearer what preconditions connect with what output

Comment (kripken) [reply tests/test_other.py]: Hmm, there are 96 combinations, though, so flattening it all out seems unpleasant. But I'll try to flatten out those if-else chains, those do seem ugly I agree.

File: tests/test_other.py (hunk 3 above)
Comment (kripken) [reply tests/test_other.py]: Done.

File: tests/test_other.py (hunk 4 above)
Comment (kripken) [reply tests/test_other.py]: Flattened out the ifs and added clarifying comments.

Comment (jgravelle-google) [reply tests/test_other.py]: Oh I missed the outer two loops, I thought it was only 16 total which seemed less bad. In theory one could also only test a subset to reduce the cases further, though that trades off against completeness.

Regardless, the flattening (and clarifying comments) works even better, so let's keep that.

File: cmake/Modules/Platform/Emscripten.cmake [hunk 5]
```diff
@@ -46,7 +46,10 @@ endif()
 
 # In order for check_function_exists() detection to work, we must signal it to pass an additional flag, which causes the compilation
 # to abort if linking results in any undefined symbols. The CMake detection mechanism depends on the undefined symbol error to be raised.
-set(CMAKE_REQUIRED_FLAGS "-s ERROR_ON_UNDEFINED_SYMBOLS=1")
+# Disable wasm in cmake checks so that (1) we do not depend on wasm support just for configuration (perhaps the user does not intend
+# to build to wasm; using asm.js only depends on js which we need anyhow), and (2) we don't have issues with a separate .wasm file
+# on the side, async startup, etc..
+set(CMAKE_REQUIRED_FLAGS "-s ERROR_ON_UNDEFINED_SYMBOLS=1 -s WASM=0")
```
Comment (juj) [inline cmake/Modules/Platform/Emscripten.cmake]: This looks good for now, although in the future we will definitely want to remove this when wasm-capable Node.jses are widely available - we do already pass `-DCMAKE_CROSSCOMPILING_EMULATOR` to point to Node.js, so as long as user has a wasm-capable Node.js I believe it should just work to set `-s WASM=1` (?)

File: src/postamble.js [hunk 6]
```diff
@@ -187,7 +187,7 @@ Module['callMain'] = function callMain(args) {
   HEAP32[(argv >> 2) + argc] = 0;
 
 #if EMTERPRETIFY_ASYNC
-  var initialEmtStackTop = Module['asm']['emtStackSave']();
+  var initialEmtStackTop = Module['emtStackSave']();
```
Comment (juj) [inline src/postamble.js]: Was this change intentional? (doesn't look just like a asm.js -> wasm defaults change?)

File: src/settings.js [hunk 7]
```diff
@@ -311,8 +312,11 @@ var STB_IMAGE = 0; // Enables building of stb-image, a tiny public-domain librar
                    // When enabled, stb-image will be used automatically from IMG_Load and IMG_Load_RW. You
                    // can also call the stbi_* functions directly yourself.
 
-var LEGACY_VM_SUPPORT = 0; // Enable this to get support for non-modern browsers, node.js, etc. This adds:
-                           //  * Polyfilling for Math.clz32, Math.trunc, Math.imul, Math.fround
+var LEGACY_VM_SUPPORT = 0; // Enable this to get support for non-modern browsers, node.js, etc. This gives you
+                           // the highest possible probability of the code working everywhere, even in rare old
+                           // browsers and shell environments. Specifically:
+                           //  * Add polyfilling for Math.clz32, Math.trunc, Math.imul, Math.fround.
+                           //  * Disable WebAssembly.
```
Comment (juj) [inline src/settings.js]: Disabling WebAssembly when `-s LEGACY_VM_SUPPORT=1` makes me feel sad about the existence of the whole `LEGACY_VM_SUPPORT` setting overall, it might be on the road to becoming a bit of a "all non-latest features pile in here" type of setting.

Does `-s LEGACY_VM_SUPPORT=1 -s WASM=1` work to give one backwards compatibility while also targeting Wasm?

**Comment by juj** [comment]
lgtm

File: src/postamble.js (hunk 6 above)
Comment (kripken) [reply src/postamble.js]: It's a bugfix, that method is exported and should be accessed like any other exported method, on module - this was noticeable in wasm because of async compilation which changes when `Module['asm']` is set. Could have been in a separate PR though, now that I think of it, sorry about that.

Comment (kripken) [reply src/postamble.js]: Or actually I think this was an issue with metadce which is only in wasm (and needs proper usages, not ones on Module.asm), and not async startup.

File: src/settings.js (hunk 7 above)
Comment (kripken) [reply src/settings.js]: Good point, in the long term maybe we want to split it out. For now, though, afaik it is used by people to target things like IE11, and other really older browsers without Math.clz etc. - those definitely also don't have wasm. So the setting is meaningless with wasm enabled, currently.

**Comment by kripken** [comment]
Thanks for the reviews!

I'll tag a minor version early this week and land this afterwards.

**Comment by kripken** [comment]
In the past we considered a major version bump for a change like this. But when we got to process details, interest seemed to evaporate. Anyone feel strongly that this should wait on that?

File: cmake/Modules/Platform/Emscripten.cmake (hunk 5 above)
Comment (curiousdannii) [reply cmake/Modules/Platform/Emscripten.cmake]: Widespread Node support is already there, except for those on v6 LTS. And those using emsdk will be given a recent enough Node version through that.

**Comment by curiousdannii** [comment]
My voice doesn't matter much, but I still think this should be majorish. Whether that's a true 2.0 or just a 1.38.

(IIRC the main reason you weren't following semver was the difficulty of getting the test servers reliably running green. IMO that shouldn't be a reason to delay versioning at any level - it doesn't matter if the bug is present in a major or a minor release, you're still going to publish a fix for it eventually. No one should be using 1.38.0 long term when 1.38.1 exists.)

**Comment by caiiiycuk** [comment]
I think this should wait while issue with slow loading time in Chrome will be solved. This is related to recent [discussion](https://groups.google.com/forum/#!topic/emscripten-discuss/Ft3UT0PHX_w) in group. For my project -s WASM=1 build is completely unusable, cause WebAssembly.instantiate in chrome tooks 6 seconds, while in asm.js mode corresponding eval takes 1.5 second. I load script manually and use eval to run it, so I think in my case eval time should be same as instantiate time but not. For sure it is chrome bug, cause FF manage WASM=1 well, and startup time is less then eval or same (near 500ms). Moreover this time is for PC, on my android phone WebAssembly.instantiate time is start from 16 sec (!sic).

My project uses cocos2d framework, and just using is enough to reproduce this problem. Test case gives not so big numbers, but wasm builds start up time is at least 2-3 times slower (200-400 ms vs 1100-1400 ms). Test case is super simple:
```
int main() {
  return 0;
}
```

:) Just build it with -s COCOS2D=3, I've use this command:
```
em++ -s WASM=1 -s USE_COCOS2D=3  --std=c++11 -O3 main.cpp -o main.js
```

And for tracking instantiate time:
```
      var instantiateImpl = WebAssembly.instantiate;
      WebAssembly.instantiate = (function(_this) {
        return function() {
          var promise, startedAt;
          startedAt = Date.now();
          promise = instantiateImpl.apply(null, arguments);
          promise.then(function() {
            Module.print('wasmInstantiateTime ' + (Date.now() - startedAt) + ' ms');
          });
          return promise;
        };
      })(this);
```

I don't feel that it is not enough to fill chrome bug, cause I guess that without any 'breaking news' they will ignore this bug report. But I think it's blocker for this PR, cause it's surprisingly that wasm is slower.

You can find ready to run test case (use main.html). Thanks!
[main.zip](https://github.com/kripken/emscripten/files/1888290/main.zip)

--
Ubuntu 16.04.1
Chrome:  65.0.3325.181

**Comment by kripken** [comment]
@curiousdannii I agree with you in spirit, a switch to wasm by default feels like it should be at least 1.38. It's just been hard to improve our testing, we have pretty poor coverage on non-linux right now, which makes me uncomfortable to declare 1.38 or merge to master.

In any case, though, I think we should merge this PR before 1.38 anyhow, just so we have some time with the change on incoming to stabilize (we'll surely find bugs!). Later on, we can merge it to master and bump to 1.38, and maybe our testing will be better by then too, wishful thinking ;)

@caiiiycuk I don't think I can reproduce the problem on chrome 64 or 66: it is slower than firefox, but just a little (0.75 vs 0.5 seconds, it reports). Or is the problem only noticeable in comparison to asm.js? Building locally, I see chrome is indeed slightly slower on wasm, but just 300ms vs 250ms for asm.js.

I'm also on Ubuntu 16.04.

I mentioned this to @dschuff who said he'd look into it too, maybe he can reproduce it?

**Comment by caiiiycuk** [comment]
I can confirm that in new beta chrome (v66) for this test initialization time is same as for asm.js and for wasm. But for real project the problem still exists 1.5 sec (asm.js) v 6 sec (wasm). Don't have new test case yet.

**Comment by kripken** [comment]
I updated the release proposal in #6174. The new proposal there makes it more practical to merge to master / bump the major version number. If that proposal is accepted, then I'd suggest we land this as follows:

 * Create 1.38.0, merge that to master.
 * Merge wasm-by-default to incoming.
 * After that stabilizes, create 1.39.0, merge that to master.

That will give us a reasonably stable version before and after this large change.

**Comment by kripken** [comment]
Ok, I think the last comment is the plan then. I'll wait a few days to see if incoming's last tag is stable (1.37.38) or if anyone has any concerns, then move forward if things look good.

**Comment by flagxor** [comment]
Is this moving forward?

**Comment by kripken** [comment]
Hit some unrelated delays due to CI problems, but should land later this week.

In more detail: As of an hour ago there is a tagged version (1.37.40) that should be good. If no new issues are found later this week, I'll merge that to master, and then merge this branch to incoming.

**Comment by flagxor** [comment]
Thanks!
